probe_src/python/probe_py/manual/cli.py

```diff
@@ -146,5 +195,539 @@ def dump(
+            return base64.b64encode(obj).decode('ascii')
+        raise TypeError(f"Type {type(obj)} not serializable")
+    
+    # Convert the dictionary to JSON
+    return json.dumps(process_dict, default=custom_serializer, indent=4)
+
+def generate_random_pid() -> int:
+    min_pid = 1
+    max_pid = 32767
+    random_pid = random.randint(min_pid, max_pid)
+    return random_pid
+
+def upload_files(cmd:list[str])->None:
+    try:
+        subprocess.run(cmd, check=True, capture_output=True, text=True)
+        typer.echo("File transfer successful.")
+    except subprocess.CalledProcessError as e:
+        # Capture and print the error message
+        typer.echo(f"Error occurred during file transfer: {e}")
+        typer.echo(f"Exit code: {e.returncode}")
+        typer.echo(f"Error output: {e.stderr}")
+
+
+def check_and_create_remote_file(hostname:str, port:int, username:str, remote_file_path:pathlib.Path) -> None:
+    try:
+        # Check if the file exists on the remote server
+        check_command = f"ssh -p {port} {username}@{hostname} 'test -f {remote_file_path} && echo \"File exists\" || echo \"File does not exist\"'"
+        result = subprocess.run(check_command, shell=True, check=True, capture_output=True, text=True)
+        output = result.stdout.strip()
+        if output == "File does not exist":
+            # Create the file if it does not exist
+            create_command = f"ssh -p {port} {username}@{hostname} 'touch {remote_file_path}'"
+            subprocess.run(create_command, shell=True, check=True)
+    except subprocess.CalledProcessError as e:
+        print(f"Command failed with exit code {e.returncode}")
+        print(f"Error output: {e.stderr}")
+    except Exception as e:
+        print(f"An error occurred: {e}")
+
+
+def get_remote_home(remote_user:str, remote_host: str, port: int)->pathlib.Path:
+    remote_xdg = get_remote_xdg_data_home(remote_user, remote_host, port)
+    if remote_xdg == "":
+        home = get_remote_home_env(remote_user, remote_host, port)
+        remote_home = pathlib.Path(f"{home}/.local/share") / "PROBE"
+    else:
+        remote_home = pathlib.Path(f"home/{remote_xdg}/.local/share") / "PROBE"
+    return remote_home
+
+def get_remote_xdg_data_home(remote_user: str, remote_host: str, port: int)-> str | None:
+    try:
+        ssh_command = [
+            "ssh",
+            "-p",
+            str(port),  # Specify the port
+            f"{remote_user}@{remote_host}",  # Remote user and host
+            "echo $XDG_DATA_HOME",  # Command to retrieve XDG_DATA_HOME
+        ]
+        result = subprocess.run(ssh_command, capture_output=True, text=True, check=True)
+        xdg_data_home = result.stdout.strip()
+        return xdg_data_home
+    except subprocess.CalledProcessError as e:
+        print(f"Error retrieving XDG_DATA_HOME: {e}")
+        return None
+
+def get_remote_home_env(remote_user: str, remote_host: str, port: int)->str | None:
```

Comment:
```suggestion
 @functools.lru_cache(maxsize=128)
def get_remote_home_env(remote_user: str, remote_host: str, port: int)->str | None:
```

This prevents the home from being "computed" every time, for hosts we've already seen in the lifetime of the program.